### PR TITLE
refactor: refactor alter table procedure

### LINKED ISSUE #3637: Alter table procedure should fetch table info during the execution
### What type of enhancement is this?

Tech debt reduction

### What does the enhancement do?

Now, submitting an altering table procedure requires fetching the table info before the procedure acquires locks(before the execution), which may cause the procedure to fail due to using the staled table info. (other procedures may acquire locks and modify the table info)

https://github.com/GreptimeTeam/greptimedb/blob/86d377d028d592764109e4622ebd17bfc7710be4/src/common/meta/src/ddl/alter_table.rs#L63-L71

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
close #3637
## What's changed and what's your intention?

1. Fetch the table info during the execution.
2. Refactor alter table procedure
3. Add mores tests

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/common/meta/src/ddl/alter_table.rs [hunk 1]
```diff
@@ -343,12 +202,13 @@ impl AlterTableProcedure {
     fn lock_key_inner(&self) -> Vec<StringKey> {
         let mut lock_key = vec![];
         let table_ref = self.data.table_ref();
+        // Safety: checked before.
```
Comment (MichaelScofield) [inline src/common/meta/src/ddl/alter_table.rs:205]: Why is retrieving table id may be considered "unsafe" here?

File: src/common/meta/src/ddl/tests/alter_table.rs [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+};
+use crate::error::Result;
+use crate::key::table_name::TableNameKey;
+use crate::key::table_route::TableRouteValue;
+use crate::peer::Peer;
+use crate::rpc::ddl::{AlterTableTask, CreateTableTask};
+use crate::rpc::router::{Region, RegionRoute};
+use crate::test_util::{new_ddl_context, MockDatanodeHandler, MockDatanodeManager};
+
+pub fn test_create_table_task(name: &str, table_id: TableId) -> CreateTableTask {
```
Comment (fengjiachun) [inline src/common/meta/src/ddl/tests/alter_table.rs:50]: `Tests` looks good to me!

**Comment by WenyXu** [comment]
@MichaelScofield @fengjiachun PTAL

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**